torch/cuda/amp/autocast_mode.py

```diff
@@ -1,146 +1,14 @@
-
-    Type mismatch errors *in* an autocast-enabled region are a bug; if this is what you observe,
-    please file an issue.
-
-    ``autocast(enabled=False)`` subregions can be nested in autocast-enabled regions.
-    Locally disabling autocast can be useful, for example, if you want to force a subregion
-    to run in a particular ``dtype``.  Disabling autocast gives you explicit control over
-    the execution type.  In the subregion, inputs from the surrounding region
-    should be cast to ``dtype`` before use::
-
-        # Creates some tensors in default dtype (here assumed to be float32)
-        a_float32 = torch.rand((8, 8), device="cuda")
-        b_float32 = torch.rand((8, 8), device="cuda")
-        c_float32 = torch.rand((8, 8), device="cuda")
-        d_float32 = torch.rand((8, 8), device="cuda")
-
-        with autocast():
-            e_float16 = torch.mm(a_float32, b_float32)
-
-            with autocast(enabled=False):
-                # Calls e_float16.float() to ensure float32 execution
-                # (necessary because e_float16 was created in an autocasted region)
-                f_float32 = torch.mm(c_float32, e_float16.float())
-
-            # No manual casts are required when re-entering the autocast-enabled region.
-            # torch.mm again runs in float16 and produces float16 output, regardless of input types.
-            g_float16 = torch.mm(d_float32, f_float32)
-
-    The autocast state is thread-local.  If you want it enabled in a new thread, the context manager or decorator
-    must be invoked in that thread.  This affects :class:`torch.nn.DataParallel` and
-    :class:`torch.nn.parallel.DistributedDataParallel` when used with more than one GPU per process
-    (see :ref:`Working with Multiple GPUs<amp-multigpu>`).
-
-    Args:
-        enabled(bool, optional, default=True):  Whether autocasting should be enabled in the region.
-    """
-    def __init__(self, enabled=True):
-        if enabled and amp_definitely_not_available():
-            warnings.warn("torch.cuda.amp.autocast only affects CUDA ops, but CUDA is not available.  Disabling.")
-            self._enabled = False
-        else:
-            self._enabled = enabled
-
-    def __enter__(self):
-        self.prev = torch.is_autocast_enabled()
-        torch.set_autocast_enabled(self._enabled)
-        torch.autocast_increment_nesting()
-
-    def __exit__(self, *args):
-        # Drop the cache when we exit to a nesting level that's outside any instance of autocast.
-        if torch.autocast_decrement_nesting() == 0:
-            torch.clear_autocast_cache()
-        torch.set_autocast_enabled(self.prev)
-        return False
-
-    def __call__(self, func):
-        @functools.wraps(func)
-        def decorate_autocast(*args, **kwargs):
-            with self:
-                return func(*args, **kwargs)
-        return decorate_autocast
 
+autocast = functools.partial(torch.autocast, device_type='cuda', fast_dtype=torch.float16)
```

Comment:
this enables surprising behavior where `torch.cuda.amp.autocast(device_type='cpu') is valid. Instead you can have `torch.cuda.amp.autocast` inherit from `torch.autocast`, and have its `init` not accept `device_type` argument.

```
class autocast(torch.autocast):
    def __init__(self, fast_dtype=torch.float16):
          super().__init__(device_type="cuda", fast_dtype=fast_dtype)
```
You can also put a check here that bfloat16 is not called on an architecture before ampere, and documentation link pointing to torch.autocast (because we still want torch.cuda.amp.autocast to be documented).